fs: Fix null pointer exception caused by async fs_unmount

cdesjardins opened the pull request:
If fs_unmount is invoked while fs i/o is ongoing then it will cause a NULL pointer exception due to the unchecked and unprotected dereference of the fs pointer.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/73501

pdgendt commented on subsys/fs/fs.c:
```diff
@@ -130,7 +130,7 @@ static int fs_get_mnt_point(struct fs_mount_t **mnt_pntp,
 
 /* File operations */
-int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
+static int _fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
```
Function names should avoid starting with underscores IMO. Maybe `fs_open_locked` shows the intent.

Also maybe add `ALWAYS_INLINE` to avoid the additional stack frame.

de-nordic commented:
This should be configurable by user as it is up to user setup, mostly, where mounted fs is used by more than one thread.
Users that do not have the issue should be able to turn off the feature as it adds code and impacts execution paths.

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -336,13 +407,8 @@ int fs_opendir(struct fs_dir_t *zdp, const char *abs_path)
 
 	if (strcmp(abs_path, "/") == 0) {
 		/* Open VFS root dir, marked by zdp->mp == NULL */
-		k_mutex_lock(&mutex, K_FOREVER);
```
This is not correct mutext to use. This one is used for protecting mount points so that when path to FS is figured out the mount list do not get removed/added.
Using this mutex for ops like fs_open means that you are globally locking VFS to perform operation of some mount point, so if you take a lot of time to write to one mount point the other thread has to wait on own fs_ op, even if that is applied to different mount point.

cdesjardins committed:
fs: Fix null pointer exception caused by async fs_unmount

The mp pointer is in fs_file_t and fs_dir_t so if the fs pointer is made
NULL then subsequent file I/O operations will cause a NULL pointer
exception. Removing the mount point from the list is threadsafe and
should be sufficient.

Signed-off-by: Chris Desjardins <chris@arch-embedded.com>

cdesjardins commented:
Thanks for your comments, and indeed I believe there is a much simpler solution which I have tested on my platform and pushed.

de-nordic commented:
> Thanks for your comments, and indeed I believe there is a much simpler solution which I have tested on my platform and pushed.

The issue you have logged is still correc.

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -791,9 +791,6 @@ int fs_unmount(struct fs_mount_t *mp)
 		goto unmount_err;
 	}
 
-	/* clear file system interface */
```
I understand that this fixes problem with NULL pointer exception but unfortunately now systems continue write to system that is no longer valid.

cdesjardins replied:
In my testing the i/o errors out and returns an error code, which is a huge improvement over the NULL pointer exception.

de-nordic replied:
Remove the "Fixes https://github.com/zephyrproject-rtos/zephyr/issues/73501" from commit and description as this does not fix the problem, it only makes it manifest in a less violent way.

cdesjardins replied:
But it does fix "the problem", which is NULL pointer dereference caused by a user removing a usb drive during file i/o.

bas-archembedded replied:
@de-nordic  Please check https://github.com/zephyrproject-rtos/zephyr/pull/78018 as an alternative.

MaureenHelm commented:
@de-nordic please revisit

de-nordic reviewed (approved):
OK, let it be.
It still does not make the FS thread safe, but reduces fallout from unmounting system somebody else is using.

A proper way would be to not allow unmounting system being used by somebody else
I think that https://github.com/zephyrproject-rtos/zephyr/pull/78018 may deliver that, once completed.
